tests/trainer/test_dataloaders.py

```diff
@@ -1020,83 +923,74 @@ def train_dataloader(self):
+@pytest.mark.parametrize("dataloader_wrapper", [CustomNotImplementedErrorDataloader, CustomInfDataloader])
+def test_train_dataloader_not_implemented_error(tmpdir, check_interval, dataloader_wrapper):
     """Test not_implemented_error train data loader (e.g. IterableDataset)"""
 
-    model = EvalModelTemplate()
-    model.train_dataloader = model.train_dataloader__not_implemented_error
-    model.val_dataloader = model.val_dataloader__not_implemented_error
+    class CustomBoringModel(BoringModel):
+        def train_dataloader(self):
+            return dataloader_wrapper(DataLoader(RandomDataset(32, 64)))
 
+        def val_dataloader(self):
+            return dataloader_wrapper(DataLoader(RandomDataset(32, 64)))
+
+    model = CustomBoringModel()
     trainer = Trainer(default_root_dir=tmpdir, max_steps=5, max_epochs=1, val_check_interval=check_interval)
     trainer.fit(model)
     # verify training completed
     assert trainer.state.finished, f"Training failed with {trainer.state}"
 
 
-def test_train_dataloader_not_implemented_error_failed(tmpdir):
-    """Test not_implemented_error train data loader (e.g. IterableDataset)"""
-    model = EvalModelTemplate()
-    model.train_dataloader = model.train_dataloader__not_implemented_error
-
-    trainer = Trainer(default_root_dir=tmpdir, max_steps=5, max_epochs=1, val_check_interval=0.5)
-
-    with pytest.raises(MisconfigurationException, match="using an IterableDataset"):
-        trainer.fit(model)
-
-
-def test_val_dataloader_not_implemented_error_failed(tmpdir):
-    """Test not_implemented_error train data loader (e.g. IterableDataset)"""
-    model = EvalModelTemplate()
-    model.val_dataloader = model.val_dataloader__not_implemented_error
-
-    trainer = Trainer(default_root_dir=tmpdir, max_steps=5, max_epochs=1, limit_val_batches=0.5)
+@pytest.mark.parametrize(
+    "stage", [RunningStage.TRAINING, RunningStage.VALIDATING, RunningStage.TESTING, RunningStage.PREDICTING]
+)
+@pytest.mark.parametrize("dataloader_wrapper", [CustomNotImplementedErrorDataloader, CustomInfDataloader])
+def test_inf_dataloader_raise_error_with_partial_batch_limits(tmpdir, stage, dataloader_wrapper):
+    """Test limit_batch error with inf dataloader (e.g. IterableDataset)"""
+    model = BoringModel()
+    setattr(
+        model, f"{stage.dataloader_prefix}_dataloader", lambda: dataloader_wrapper(DataLoader(RandomDataset(32, 64)))
+    )
+    trainer_kwargs = {"default_root_dir": tmpdir, "max_epochs": 1, f"limit_{stage.dataloader_prefix}_batches": 0.5}
+    trainer = Trainer(**trainer_kwargs)
+    trainer_fn = "fit" if stage == RunningStage.TRAINING else stage.value
 
-    with pytest.raises(MisconfigurationException, match="using an IterableDataset"):
+    with pytest.raises(MisconfigurationException, match=r"using an IterableDataset .* must be `0.0`, `1.0`"):
+        getattr(trainer, trainer_fn)(model)
         trainer.fit(model)
```

Comment:
there are two trainer calls here, but the raised error can only occur in one of them.

Reply:
good catch! forgot to remove it